decomposer.go

```diff
@@ -69,7 +71,19 @@ func (d *Decimal) Decompose(buf []byte) (form byte, negative bool, coefficient [
 	// Finite form.
 	negative = d.Negative
 	exponent = d.Exponent
-	coefficient = d.Coeff.Bytes()
+
+	sizeInBytes := (d.Coeff.BitLen() + 8 - 1) / 8 // math.Ceil(d.Coeff.BitLen()/8.0)
+	if cap(buf)-len(buf) >= sizeInBytes {
```

Comment:
> cap(buf)-len(buf)

What does it mean for `len(buf) != 0`? Should we preserve the data in the slice and only use the unused capacity? The contract here, which comes from https://github.com/golang/go/issues/30870, is ambiguous.

The usual convention for when a function like this takes a scratch space buffer is that it is permitted to overwrite any data in the buffer. So the function checks the capacity (`if cap(buf) >= sizeInBytes`), sets the length to the needed length if there's room (`buf = buf[:sizeInBytes]`), and then writes into the buffer from the beginning (`coefficient = d.Coeff.FillBytes(buf)`).

Reply:
Yeah that makes sense, I will change to this approach instead.

Reply:
Since you made that change, I don't understand why the `buf = append(buf, make([]byte, sizeInBytes-len(buf))...)` logic is needed, instead of just `buf = buf[:sizeInBytes]`.

Reply:
Yes you're right that's not necessary. I actually learned today that it's allowed to increase the length if the capacity is big enough without append.